### ISSUE TITLE
No updates since July, website stuck?

### ISSUE THREAD

**Issue by leahneukirchen** [schemeorg-community/planet.scheme.org#8]
planet.scheme.org hasn't been updated since July 24, but a few blogs had posts in the meantime, does the server need a restart?

**Comment by lassik** [comment]
The server is fine but the blog aggregator, Planet Venus, is held together with duct tape and bubble gum. It keeps breaking in various ways. In case you have spare cycles to try to fix it, I can give SSH access to the box. Beware that the work is not fun.

**Comment by leahneukirchen** [comment]
Uh, I'll give it a try, my key is at https://github.com/leahneukirchen.keys

**Comment by lassik** [comment]
Thanks. Please try `ssh leahneukirchen@tuonela.scheme.org`

My bash history says that I last tried the following command line:

`cd /production/planet/ && sudo -u prod-planet tar -cf - cache | docker run --rm -i prod-planet | sudo -u prod-planet tar -xvf -`

and the container stopped with the following error:

```
Downloading https://i.snap.as/hrje8kg1.png
Saving into images/cache/9f529e65913c774bac185bac4bffdc58acc7a761.png
Downloading https://i.snap.as/vp7SwJDz.png
Saving into images/cache/30fb7e963a45d1d6e9adbc7dbe8f6184ac93bc37.png
*** SYSTEM-ERROR: write failed on #<oport /planet/images-cache.scm 0x7f99e771a900>: Bad file descriptor
    While loading "./../images.scm" at line 91
```

There have been at least two different ways to have this contraption run as a cron job, and there were problems with each. There have also been many other, non-cron-related problems with the aggregator that I've mercifully forgotten.

**Comment by lassik** [comment]
The above error comes from [`images.scm`](https://github.com/schemeorg-community/planet.scheme.org/blob/master/images.scm). It's a custom script I wrote that runs using Gauche after Planet Venus has done its job. `images.scm` downloads all the images from the blog posts into a local cache since Planet Venus doesn't do that.

**Comment by leahneukirchen** [comment]
I'll have a look, thanks!

**Comment by lassik** [comment]
Thank you for your service. I wasn't kidding about the duct tape.

The simplest "patch on top of a patch" fix is probably to comment out the shell command that runs `images.scm`. In that case, the web page generated by Planet Venus ought to use off-site URLs in the `<img ...>` tags. However, there's no clear reason why `images.scm` should encounter `Bad file descriptor`. I hope it's not some obscure feature of Docker.

**Comment by leahneukirchen** [comment]
So when I ran it, it worked fine... One post has mochibake but that seems unrelated.

**Comment by lassik** [comment]
lol. It works in mysterious ways.

It's good to know that it's not systematically broken. We've been reluctant to spend time on it since the architecture is pretty much broken beyond repair. It would be best to write a new blog aggregator in Scheme or Racket, but no one has had the time to do it yet.

**Comment by leahneukirchen** [comment]
The mojibake is a result of this imo weird behavior of feedparser:

```
    But, if
    the HTTP Content-Type is text/xml, text/*+xml, or
    text/xml-external-parsed-entity, the encoding given in the XML prefix
    within the document is ALWAYS IGNORED and only the encoding given in
    the charset parameter of the HTTP Content-Type header should be
    respected, and it defaults to 'us-ascii' if not specified.
```

**Comment by leahneukirchen** [comment]
Where is the image-cache supposed to be stored anyway?

**Comment by lassik** [comment]
> The mojibake is a result of this imo weird behavior of feedparser:

Good to know. Potentially relevant to #5.

> Where is the image-cache supposed to be stored anyway?

As regular files under `/production/planet/www/images/cache/`.

**Comment by lassik** [comment]
https://planet.scheme.org/ is simply the contents of `/production/planet/www/` served as a static site by nginx.

**Comment by leahneukirchen** [comment]
Those are the images, but I mean the cache file that script writes.

We could try updating Gauche, perhaps there was a bug in it.

**Comment by leahneukirchen** [comment]
Also, how was this meant to be executed periodically? The cronjob in /etc/cron.d doesn't exist.

**Comment by lassik** [comment]
> Those are the images, but I mean the cache file that script writes.

My mistake. `images-cache.scm` is not shipped out of the container. It should be part of the tarball.

The culprit is probably this line In `planet-container.sh`:

`tar --sort name -cf - www >&3`

should be 

`tar --sort name -cf - images-cache.scm www >&3`

> We could try updating Gauche, perhaps there was a bug in it.

No idea.

> Also, how was this meant to be executed periodically? The cronjob in /etc/cron.d doesn't exist.

```
$ sudo find /etc/ -iname '*planet*'
/etc/cron.d/scheme-prod-planet
```

**Comment by leahneukirchen** [comment]
```
$ cat /etc/cron.d/scheme-prod-planet
0 0-23/6 * * * prod-planet /production/planet/planet/planet.sh
$ cat /production/planet/planet/planet.sh
cat: /production/planet/planet/planet.sh: No such file or directory
```

**Comment by lassik** [comment]
Yeah, I'm not surprised that it's broken. It has been variously working and not working in the past.

The wider context here is that we should standardize how static sites under scheme.org are generated. The Planet counts as a static site for this purpose. The only twist is that unlike most static sites, it is rebuilt on a schedule.

I think the best approach for such a standard is a container that outputs a tar file. The Planet has taken a step in this direction, but since we're all busy and passionate sysadmins are hard to find around Lisp, the job is not done.

**Comment by leahneukirchen** [comment]
I'm not sure why you prefer these tar tricks over docker volumes, but ymmv.

I can rewrite that script to use docker, then we can run it on a schedule again?

**Comment by lassik** [comment]
Sure. That's a good fix.

**Comment by leahneukirchen** [comment]
But user `prod-planet` can't use docker directly...

**Comment by lassik** [comment]
> The Planet counts as a static site for this purpose. The only twist is that unlike most static sites, it is rebuilt on a schedule.

My memory is refreshing. The cache is another oddity. Most static site builds don't use a dynamic cache as input.

> But user `prod-planet` can't use docker directly...

Yeah. This is death by a thousand cuts.

I'll add `prod-planet` to the `docker` group until we think of something elegant. Feel free to use a Docker volume if you like.

**Comment by lassik** [comment]
OK.

```
$ sudo -iu prod-planet id
uid=4018(prod-planet) gid=4018(prod-planet) groups=4018(prod-planet),100(users),110(docker)
```

**Comment by leahneukirchen** [comment]
Ok, I adjusted the cron job now, it should only report when if fails, not sure who gets a mail then but I can also put myself into MAILTO. I'll watch it for a few days.

**Comment by arthurgleckler** [comment]
This is great!  Thank you very much for looking into this.  Keeping Planet Scheme up to date will make it much more useful.

**Comment by lassik** [comment]
What's up with `planet.sh` and `cron.sh` in `/production/planet`? The crontab entry in `/etc/cron.d/scheme-prod-planet` still points at `/production/planet/planet/planet.sh`.

Re: MAILTO, how does cron send mail? I don't think we have a working SMTP server on that box. IIRC Debian ships with Exim, but I can't find a running process that belongs to it.

**Comment by lassik** [comment]
The fact that we have a file called `/production/planet/planet/planet.sh` says something about the state of disrepair :)

**Comment by leahneukirchen** [comment]
Fixed the path (did i forget to save oO), seems like mail doesn't work indeed.

**Comment by lassik** [comment]
We can set up Exim. But in today's adversarial climate, does the mail have a fighting chance of landing in people's inboxes?

**Comment by leahneukirchen** [comment]
I wouldn't bother for now (and use postfix if we actually need an MTA).

**Comment by lassik** [comment]
Fair enough.

Thanks a lot for taking on this chore. I've been begging people to help for months.

If this keeps working, let's commit `cron.sh` and the fixed crontab entry to git. The principle is that all server configuration is public - no security by obscurity.

**Comment by leahneukirchen** [comment]
The bug happened once more, but when I tried to reproduce it, it worked again...

**Comment by lassik** [comment]
Do you mean the `Bad file descriptor` bug?

**Comment by leahneukirchen** [comment]
yes, that one. for mysterious reasons I still can't reproduce it manually. i'll attach a strace into the cron script...

**Comment by leahneukirchen** [comment]
Does something reset the cron job? It's back to `/production/planet/planet/planet.sh` as of October 25 (and doesn't work since then).

**Comment by lassik** [comment]
Indeed, it seems our Ansible config overwrites it with the old broken stuff. I'll fix it within an hour.

Is `cron.sh` the right script to run?

**Comment by leahneukirchen** [comment]
Exactly.

**Comment by leahneukirchen** [comment]
done

**Comment by lassik** [comment]
Some tmux processes are still up.

**Comment by leahneukirchen** [comment]
also closed now

**Comment by lassik** [comment]
Thank you.

Actually >1 hour because systemd is doing what it does best. Failing in mysterious ways that cannot be googled.

**Comment by lassik** [comment]
OK. The server should finally be back in order including the cronjob.

```
$ cat /etc/cron.d/scheme-prod-planet
0 0-23/6 * * * prod-planet /production/planet/cron.sh
```

**Comment by arthurgleckler** [comment]
Is there something that needs to be done to make the RSS feed on scheme.org match?

**Comment by lassik** [comment]
@arthurgleckler: It's still a manual process. In the [schemeorg repo](https://github.com/schemeorg/schemeorg), `scripts/www.sh && scripts/wwwproduction.sh`. The first command fetches the feed and generates the pages. The second uploads.

Eventually this should also be run as a server-side cronjob. Perhaps in Docker. A standard repo layout would be the key to this problem and several others.

**Comment by lassik** [comment]
Anyway, I just updated manually.
